ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,884 @@
+   * Checks the sanity of the provided {@code details}
+   * @param details the {@link CompactionDetails} to use for compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log or if one or more offsets
+   * in the segments to compact are in the journal.
+   */
+  private void checkSanity(CompactionDetails details) {
+    List<String> segmentsUnderCompaction = details.getLogSegmentsUnderCompaction();
+    LogSegment lastSegment = null;
+    // all segments should be available
+    for (String segmentName : segmentsUnderCompaction) {
+      lastSegment = srcLog.getSegment(segmentName);
+      if (lastSegment == null) {
+        throw new IllegalArgumentException(segmentName + " does not exist in the log");
+      }
+    }
+    // last offset to compact should be outside the range of the journal
+    Offset lastSegmentEndOffset = new Offset(lastSegment.getName(), lastSegment.getEndOffset());
+    if (lastSegmentEndOffset.compareTo(srcIndex.journal.getFirstOffset()) >= 0) {
+      throw new IllegalArgumentException("Some of the offsets provided for compaction are within the jounral");
+    }
+  }
+
+  /**
+   * Copies valid data from the source log segments into the swap spaces.
+   * <p/>
+   * Splits the compaction into two cycles if there aren't enough swap spaces and completes the copy for the current
+   * cycle.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  private void copy() throws InterruptedException, IOException, StoreException {
+    setupState();
+    List<String> logSegmentsUnderCompaction = compactionLog.getCompactionDetails().getLogSegmentsUnderCompaction();
+    Offset startOffsetOfLastIndexSegmentForDelete = getStartOffsetOfLastIndexSegmentForDelete();
+    FileSpan duplicateSearchSpan = null;
+    if (compactionLog.getCurrentIdx() > 0) {
+      // only the very first log segment in the cycle can contain duplicate records.
+      LogSegment logSegment = srcLog.getSegment(logSegmentsUnderCompaction.get(0));
+      LogSegment prevSegment = srcLog.getPrevSegment(logSegment);
+      if (prevSegment != null) {
+        // duplicate data, if it exists, can only be in the previous log segment.
+        Offset startOffset = new Offset(prevSegment.getName(), prevSegment.getStartOffset());
+        Offset endOffset = new Offset(prevSegment.getName(), prevSegment.getEndOffset());
+        duplicateSearchSpan = new FileSpan(startOffset, endOffset);
+      }
+    }
+
+    for (String logSegmentName : logSegmentsUnderCompaction) {
+      LogSegment srcLogSegment = srcLog.getSegment(logSegmentName);
+      Offset logSegmentEndOffset = new Offset(srcLogSegment.getName(), srcLogSegment.getEndOffset());
+      if (needsCopying(logSegmentEndOffset) && !copyDataByLogSegment(srcLogSegment, duplicateSearchSpan,
```

Comment:
I would like to keep it separate because:
1) I think it gives the sense of "if it needs to be copied, then copy"
2) More importantly, since `copyDataByLogSegment()` returns a `boolean`, I find it more understandable if it has been determined that some copying is required and I received `true/false` to indicate whether that copying finished.

If it not unintuitive this way, I strongly prefer this.